owlkettle/adw.nim

```diff
@@ -867,6 +870,157 @@ when AdwVersion >= (1, 2) or defined(owlkettleDocs):
   
   export AboutWindow
 
+## Adw.Toast
+type Toast* = ref object
+  title*: string
+  customTitle*: Widget
+  buttonLabel*: string
+  priority*: ToastPriority
+  timeout*: int
+  dismissalHandler: proc(toast: Toast)
+  clickedHandler: proc()
+  useMarkup: bool
+
+proc newToast*(
+  title: string,
+  buttonLabel: string = "", 
+  priority: ToastPriority = ToastPriorityNormal, 
+  dismissalHandler: proc(toast: Toast) = nil, 
+  clickedHandler: proc() = nil,
+  timeout: int = 5, 
+  useMarkup: bool = false,
+  customTitle: Widget = nil.Widget
+): Toast =
+  result = Toast(
+    title: title, 
+    buttonLabel: buttonLabel,
+    priority: priority,
+    timeout: timeout,
+    dismissalHandler: dismissalHandler
+  )
+  
+  when AdwVersion >= (1, 2):
+    result.clickedHandler = clickedHandler
+    result.customTitle = customTitle
+  else:
+    let isUsingCustomTitle = not customTitle.isNil()
+    if isUsingCustomTitle:
+      raise newException(LibraryError, "The customTitle field on Toast is not available when compiling for Adwaita versions below 1.2. Compile for Adwaita version 1.2 or higher with -d:adwminor=2 to enable it")
+
+    let isUsingClickedHandler = not clickedHandler.isNil()
+    if isUsingClickedHandler:
+      raise newException(LibraryError, "The clickedHandler field on Toast is not available when compiling for Adwaita versions below 1.2. Compile for Adwaita version 1.2 or higher with -d:adwminor=2 to enable it")
+      
+  when AdwVersion >= (1, 4):
+    result.useMarkup = useMarkup
+  else:
+    let isUsingUseMarkup = useMarkup == true
+    if isUsingUseMarkup:
+      raise newException(LibraryError, "The useMarkup on Toast field is not available when compiling for Adwaita versions below 1.4. Compile for Adwaita version 1.4 or higher with -d:adwminor=4 to enable it")
+
+proc toOwl(adwToast: AdwToast): Toast =
+  when AdwVersion >= (1, 4):
```

Comment:
This is solely necessary for the dismissal-event aka the signal that gets fired when a toast is dismissed.
The user gets to access the Toast that was dismissed (or rather a copy of the data) instead of an AdwToast, therefore the conversion is needed.

Note that I very intentionally here *only* forward the data, I do *not* forward the event handlers themselves because I don't think I can and I don't think we lose anything by not doing so.

Reply:
I don't think that passing the Toast back to the user when it is dismissed is actually necessary. If you need the dismissed toast, you can just access it via the closure of the event handler, can't you?

Reply:
You could make an argument that by passing a Toast you could make accessing the toast more convenient by doing so. But off the cuff, syntactically, yeah you could just do it with a closure.
As in, first define toast, then define dismissal-handler proc which captures the toast object, then assign the dismissalhandler to that toast object.

Reply:
I think I prefer just using a closure. Especially since the toast returned by toOwl will be a different object and does not contain the callbacks.

Reply:
Alrighty, done